gapic/src/main/com/google/gapid/models/AtomStream.java

```diff
@@ -301,58 +304,58 @@ public Atom getLastSelectedAtom() {
   }
   */
 
-  public TypedObservation[] getObservations(AtomIndex index) {
-    /*
-    Atom atom = getAtom(index);
-    if (atom.getObservationCount() == 0) {
-      return NO_OBSERVATIONS;
-    }
-
-    Observations obs = atom.getObservations();
-    TypedObservation[] result = new TypedObservation[obs.getReads().length + obs.getWrites().length];
-    for (int i = 0; i < obs.getReads().length; i++) {
-      result[i] = new TypedObservation(index, true, obs.getReads()[i]);
-    }
-    for (int i = 0, j = obs.getReads().length; i < obs.getWrites().length; i++, j++) {
-      result[j] = new TypedObservation(index, false, obs.getWrites()[i]);
-    }
-    return result;
-    */
-    return NO_OBSERVATIONS;
+  public ListenableFuture<Observation[]> getObservations(AtomIndex index) {
+    return Futures.transform(client.get(observationsAfter(index, Application_VALUE)), v -> {
+      Service.Memory mem = v.getMemory();
+      Observation[] obs = new Observation[mem.getReadsCount() + mem.getWritesCount()];
+      int idx = 0;
+      for (Service.MemoryRange read : mem.getReadsList()) {
+        obs[idx++] = new Observation(index, true, read);
+      }
+      for (Service.MemoryRange write : mem.getWritesList()) {
+        obs[idx++] = new Observation(index, false, write);
+      }
+      return obs;
+    });
   }
 
-  private static final TypedObservation[] NO_OBSERVATIONS = new TypedObservation[0];
-
   /**
    * Read or write memory observation at a specific command.
    */
-  public static class TypedObservation {
-    public static final TypedObservation NULL_OBSERVATION = new TypedObservation(null, false) {
+  public static class Observation {
```

Comment:
Hah! The name "TypedObservation" was a sad work around for the fact that we had a codergen'ed class named Observation and the "type" in the name referred to the fact that this class knew whether it was a read or write, while the other one didn't.

Filed a bug (https://github.com/google/gapid/issues/350) anyway, though, so the memory view can switch types automatically...